src/wallet/wallet.cpp

```diff
@@ -1892,30 +1891,27 @@ void CWallet::ReacceptWalletTransactions(interfaces::Chain::Lock& locked_chain)
     // Try to add wallet transactions to memory pool
     for (const std::pair<const int64_t, CWalletTx*>& item : mapSorted) {
         CWalletTx& wtx = *(item.second);
-        CValidationState state;
-        wtx.AcceptToMemoryPool(locked_chain, state);
+        std::string err_string;
+        wtx.SubmitMemoryPoolAndRelay(err_string);
     }
 }
 
-bool CWalletTx::RelayWalletTransaction(interfaces::Chain::Lock& locked_chain)
+bool CWalletTx::SubmitMemoryPoolAndRelay(std::string& err_string)
 {
     // Can't relay if wallet is not broadcasting
     if (!pwallet->GetBroadcastTransactions()) return false;
-    // Don't relay coinbase transactions outside blocks
-    if (IsCoinBase()) return false;
     // Don't relay abandoned transactions
     if (isAbandoned()) return false;
-    // Don't relay conflicted or already confirmed transactions
-    if (GetDepthInMainChain(locked_chain) != 0) return false;
```

Comment:
There I still have a doubt on BroadcastTransaction HaveChain check behavior. ```HaveChain = !existingCoin.IsSpent()``` doesn't mean if tx is in chain, but all of its outputs has been spent, we accept it again to mempool/relay ?

Reply:
`IsSpent()` is "misleadingly" named. Because we use `AccessCoin`: `Return a reference to Coin in the cache, or a pruned one if not found.`. Pruned in this case means all its `CTxOut` are `IsNull()`, and `IsSpent()` just calls `IsNull()`.

Reply:
Thanks @Sjors . I agree with you that `IsSpent()` is a misleading name. A more accurate name would be `DoesNotExist()`. Negating the function here determines whether the coin exists in the current UTXO set.

The check in `BroadcastTransaction()` is an optimization: if any of the outputs from the tx still exist in the UTXO set then the tx must be confirmed. If the coin has been confirmed but it's outputs don't exist, then it'll fail later (because its inputs can't exist if it has already been spent).

This check in `RelayWalletTransaction()` is also an optimization. If the transaction has been confirmed (or a conflicting transaction has been confirmed), then the block hash will be recorded in the `CMerkleTx`, and we can check in the chain whether that block exists.

An alternative to removing this call to `GetDepthInMainChain()` would be to merge https://github.com/bitcoin/bitcoin/pull/15931 first, which removes `GetDepthInMainChain`'s requirement on locking the chain.

Reply:
> Thanks @Sjors . I agree with you that `IsSpent()` is a misleading name. A more accurate name would be `DoesNotExist()`. Negating the function here determines whether the coin exists in the current UTXO set.
> 
> The check in `BroadcastTransaction()` is an optimization: if any of the outputs from the tx still exist in the UTXO set then the tx must be confirmed. If the coin has been confirmed but it's outputs don't exist, then it'll fail later (because its inputs can't exist if it has already been spent).

It might be useful to enhance the code documentation added in 976d1e485 with this information (if it isn't expected to change soon). I found it v helpful to understanding the `!existingCoin.IsSpent())` optimisation.